Causal hash/branch hash cleanup migration

ChrisPenner opened the pull request:
## Overview


There was a bug in previous versions of UCM which incorrectly used causal hashes as branch hashes. This remained undetected because there was never a need for this hash to be verifiable, and the hashes were still unique because the namespace hash was PART of the causal hash. It did however result in many identical branches being stored multiple times under different `primary_hash_id`s.

However, with the advent of Share and Sync, we now need to correctly verify these namespace hashes.

This migration fixes the issue by re-hashing namespace objects where the value_hash_id of a causal matches the self_hash_id. Luckily this doesn't change any causal hashes.

However, due to the possibility of multiple identical objects stored under different `primary_hash_id`s, we may now have multiple objects with the same `primary_hash_id`, which our DB schema doesn't allow.

To address this, we keep exactly one 'canonical' object for each hash, then remap all references to old objects into this canonical object instead. Unfortunately this requires mapping over every branch object and traversing the child references.

It was also discovered that some developers had many branches which referenced objects which weren't in their codebase. We're not yet sure how this happened, but it's unlikely to be the case for most end users, and it turned out that these references were in causals and branches which were unreachable from the root namespace. As a fix, this migration also tracks every causal and branch which is reachable from the root namespace and deletes all causals and namespaces which are unreachable. Note that this may orphan some definitions, patches, etc. which were previously referenced in an 'unreachable' branch, but they were already floating around in an unreachable state.

## Implementation

Branch hash fix:
* Adds `value_hash` to V1 Causals
* propagate the value hash and use that when creating causals.
* Splits and clarifies the types for `NamespaceHash` and `CausalHash` to prevent this mixup in the future
* Simplifies the V1 Causal type a bit.

Migration:

* Use the Sync abstraction to crawl causals in dependency in order
* For each causal, first ensure all ancestors and child branches are migrated, and their 'canonical' branch objects have been recorded.
* If we have a mapping from our value_hash_id to a corrected canonical branch_hash_id, replace the value_hash_id in place and we're done. Otherwise:
* For the branch object referenced by the value_hash_id, re-map any contained child object ids to their 'canonical' object ID and replace the blob for that object in sqlite in-place.
* if the self_hash_id and value_hash_id for the causal differ, we're done with it, otherwise:
* Re-hash the branch to obtain the correct namespace hash.
* determine whether there's an existing 'canonical' branch object with this hash as its primary_hash_id, otherwise this object is now the canonical object; save that info in the canonical object mapping.
* If there wasn't an existing canonical object, update the primary_hash_id and any hash_objects for this object to the new hash_id.
* If there IS an existing canonical object, delete the object and any matching hash_objects, any references to it will be re-mapped as the migration proceeds.
* As we crawl every branch and causal, keep a set of reachable causal_hash_id's and branch_object_ids
* After migrating everything, delete all corresponding branch objects, hash_objects, causal-parents, and causals which aren't in the set of reachable things.


This PR also adds a set of integrity checks which run after any migration, these checks:

* Assert that every causal's value_hash_id differs from its causal_hash_id
* Assert that every causal hash's value hash has an object (We can remove this check in the future if this property changes)
* Asserts that an object for every referenced child branch Obj Id exists.
* Asserts that every causal hash id referenced as a child branch has a matching causal
* Asserts that the object id and causal hash id actually refer to the same branch object for all child references.

If any of these assertions fail, the migration aborts, rolls back, and prompts the user to report a bug.

ChrisPenner committed:
Rename Branch.Hash -> Branch.CausalHash for clarity.

ChrisPenner committed:
WIP simplifying types for Causals and hashes

ChrisPenner committed:
Single type for causal hashes.

ChrisPenner committed:
Remove vestigial function

ChrisPenner committed:
Clean up some rogue hashes

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs:
```diff
@@ -274,7 +274,7 @@ sqliteCodebase debugName root localOrRemote action = do
               CodebaseOps.putRootBranch rootBranchCache (Branch.transform (Sqlite.unsafeIO . runInIO) branch1)
 
-        rootBranchUpdates :: MonadIO m => TVar (Maybe (Sqlite.DataVersion, a)) -> m (IO (), IO (Set Branch.Hash))
+        rootBranchUpdates :: MonadIO m => TVar (Maybe (Sqlite.DataVersion, a)) -> m (IO (), IO (Set Branch.CausalHash))
```
BTW this function currently does nothing, but implies that it detects root branch updates;

aryairani replied:
Yeah, we should dig in and decide whether to make it detect root branch updates, or delete it.

ChrisPenner committed:
Formatting

ChrisPenner committed:
Fix tests

ChrisPenner committed:
Add namespace hash migration

ChrisPenner committed:
Cleanup

ChrisPenner committed:
Finish writing migration for bad branches, add some sanity checks too.

ChrisPenner committed:
Keep hash_object mappings up to date.

ChrisPenner committed:
Working migration

ChrisPenner committed:
More checks

ChrisPenner committed:
Dependency-order migration WIP

ChrisPenner committed:
Finish first draft of crawled migration strategy

ChrisPenner committed:
Improve speed and reliability of migration.

ChrisPenner committed:
Better migration logging

ChrisPenner committed:
Perform integrity checks once after all migrations are done.

ChrisPenner committed:
Considerably speed up migration by only re-hashing branches which have
the same causal hash as value hash.

ChrisPenner committed:
Clarify migration docs

ChrisPenner committed:
Move 'differing causal value hash' check to integrity checks

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Branch/Type.hs:
```diff
@@ -17,15 +29,17 @@ import Unison.NameSegment (NameSegment)
 
-type Hash = Causal.RawHash Raw
+-- | A Hash for a namespace itself, it doesn't incorporate any history.
+type NamespaceHash m = Hash.HashFor (Branch0 m)
```
It's admittedly a bit silly to have the `m` here, but this was the easiest way to do this for now, and it wasn't very difficult to plumb things through 🤷🏼

aryairani replied:
Could it go away with a `forall` on the inside, or something?

ChrisPenner replied:
Unfortunately not, since it would still need to unify with the `m` in scope when you create it from the Causal,  and that concrete `m` won't match `(forall m. m)`.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal/Type.hs:
```diff
@@ -41,58 +41,60 @@ import Prelude hiding (head, read, tail)
-newtype RawHash a = RawHash {unRawHash :: Hash}
-  deriving (Eq, Ord, Generic)
+-- | Represents a hash of a causal containing values of the provided type.
+newtype CausalHash = CausalHash {unCausalHash :: Hash}
```
Concretizing the causal hash type makes things much easier to understand with no real downside. If we end up wanting to add the tag back in later we can, but it's pretty unlikely we'll ever want to do that on the v1 causal type.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Conversions.hs:
```diff
@@ -424,40 +424,30 @@ causalbranch2to1 :: Monad m => (V2.Reference -> m CT.ConstructorType) -> V2.Bran
 
 causalbranch1to2 :: forall m. Monad m => V1.Branch.Branch m -> V2.Branch.CausalBranch m
-causalbranch1to2 (V1.Branch.Branch c) = causal1to2' hash1to2cb hash1to2c branch1to2 c
+causalbranch1to2 (V1.Branch.Branch c) =
```
This function is where we were previously incorrectly using the causal hash as the branch hash.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations.hs:
```diff
@@ -68,12 +72,19 @@ ensureCodebaseIsUpToDate localOrRemote root getDeclType termBuffer declBuffer co
+          when ranMigrations $ do
+            putStrLn $ "🕵️  Checking codebase integrity..."
+            run do
+              integrityCheckFullCodebase >>= \case
```
Here's where we now run a full integrity check after any migrations.

It's pretty fast; only seems to take a few seconds at most.

ChrisPenner committed:
Bump schema version for new codebases

aryairani commented on parser-typechecker/src/Unison/Codebase/Branch.hs:
```diff
@@ -373,7 +372,7 @@ step f = runIdentity . stepM (Identity . f)
 stepM :: (Monad n, Applicative m) => (Branch0 m -> n (Branch0 m)) -> Branch m -> n (Branch m)
 stepM f = \case
-  Branch (Causal.One _h e) | e == empty0 -> Branch . Causal.one <$> f empty0
+  Branch (Causal.One _h _ e) | e == empty0 -> Branch . Causal.one <$> f empty0
```
```suggestion
  Branch (Causal.One _hc _he e) | e == empty0 -> Branch . Causal.one <$> f empty0
```

aryairani commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Branch/Dependencies.hs:
```diff
@@ -52,9 +51,9 @@ to' Dependencies {..} = Dependencies' (toList patches) (toList terms) (toList de
 fromBranch :: Applicative m => Branch m -> (Branches m, Dependencies)
 fromBranch (Branch c) = case c of
-  Causal.One _hh e -> fromBranch0 e
-  Causal.Cons _hh e (h, m) -> fromBranch0 e <> fromTails (Map.singleton h m)
-  Causal.Merge _hh e tails -> fromBranch0 e <> fromTails tails
+  Causal.One _hh _eh e -> fromBranch0 e
+  Causal.Cons _hh _eh e (h, m) -> fromBranch0 e <> fromTails (Map.singleton h m)
+  Causal.Merge _hh _eh e tails -> fromBranch0 e <> fromTails tails
   where
     fromTails m = ([(h, Branch <$> mc) | (h, mc) <- Map.toList m], mempty)
```
Do we want to do anything more with _eh here?

ChrisPenner replied:
I took a look at where this function is used in the sync algorithm, and no I don't think so; the sync algorithm syncs the branch hash ID as part of syncing the causal hash, so I don't think adding it here provides any value.

aryairani commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs:
```diff
@@ -351,7 +351,7 @@ getRootBranch doGetDeclType rootBranchCache =
         else do
           newRootHash <- Ops.expectRootCausalHash
-          if Branch.headHash b == Cv.branchHash2to1 newRootHash
+          if Branch.headHash b == Cv.causalHash2to1 newRootHash
```
What was this doing before?

ChrisPenner committed:
PR cleanup

ChrisPenner committed:
Remove commented out migration

ChrisPenner committed:
PR feedback cleanup

aryairani commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema3To4.hs:
```diff
@@ -0,0 +1,325 @@
+        dbBranch & DBBranch.childrenHashes_ %%~ \(ids@(_childBranchObjId, childCausalHashId)) -> do
+          case Map.lookup childCausalHashId canonicalBranchForCausalMap of
+            Nothing -> (([childCausalHashId], Any False), ids)
+            Just (_, canonicalObjId) -> (([], Any True), (canonicalObjId, childCausalHashId))
```
```suggestion
            Just (_, canonicalObjId) -> (([], Any (canonicalObjId /= _childBranchObjId)), (canonicalObjId, childCausalHashId))
```

ChrisPenner committed:
Add integrity check for hash_objects.

ChrisPenner committed:
Correctly set 'changes' key

aryairani reviewed (approved):
Great! Approved pending whatever changes we just discussed.

ChrisPenner committed:
Don't check hash object integrity in migrations

Some codebases (like mine) are borked in this area and we'll need to fix
them later on.

ChrisPenner committed:
Correct final migration number

ChrisPenner committed:
Check namespace hash even if branch has no children

ChrisPenner committed:
Rehash every branch regardless

ChrisPenner committed:
Use v4 of base.
